File: MATLAB/example_usage.m
% example_usage.m
% Example usage of the Gradient Descent algorithm for a multivariable function in MATLAB

% Define the function f(x, y) = (x - 3)^2 + (y + 1)^2
func = @(v) (v(1) - 3)^2 + (v(2) + 1)^2;

% Define the gradient of the function: ∇f(x, y) = [2*(x - 3), 2*(y + 1)]
grad_func = @(v) [2 * (v(1) - 3); 2 * (v(2) + 1)];

% Gradient descent parameters
initial_guess = [0; 0];
learning_rate = 0.1;
max_iterations = 1000;
tolerance = 1e-6;

% Run gradient descent
[minimum, iterations, converged] = gradient_descent(func, grad_func, initial_guess, learning_rate, max_iterations, tolerance);

% Display the results
if converged
    fprintf('Minimum found at: (%.6f, %.6f) - Iterations: %d\n', minimum(1), minimum(2), iterations);
else
    fprintf('Solution did not converge within the given tolerance and iteration limit.\n');
end

File: Ruby/example_usage.rb
# example_usage.rb
# Example usage of the Gradient Descent algorithm for a multivariable function in Ruby

require_relative 'gradient_descent'

# Örnek fonksiyon f(x, y) = (x - 3)^2 + (y + 1)^2
func = lambda { |v| (v[0] - 3)**2 + (v[1] + 1)**2 }

# Fonksiyonun gradyanı: ∇f(x, y) = [2*(x - 3), 2*(y + 1)]
gradient = lambda { |v| [2 * (v[0] - 3), 2 * (v[1] + 1)] }

# Gradient Descent parametreleri
initial_guess = [0.0, 0.0]
learning_rate = 0.1
max_iterations = 1000
tolerance = 1e-6

# GradientDescent sınıfını oluştur ve minimum bul
gd = GradientDescent.new(
  func: func,
  gradient: gradient,
  learning_rate: learning_rate,
  max_iterations: max_iterations,
  tolerance: tolerance
)

result = gd.find_minimum(initial_guess)

# Sonucu göster
if result[:converged]
  puts "Minimum bulunduğu nokta: (#{result[:minimum][0].round(6)}, #{result[:minimum][1].round(6)}) - Iterations: #{result[:iterations]}"
else
  puts "Çözüm, belirtilen tolerans ve iterasyon sınırında yakınsamadı."
end
